### PR TITLE
Provide RDP capability and (automatic) port mapping for Windows guests

### LINKED ISSUE #116: Enable RDP support for Windows guests
Add 'RDP capability' and automatic port forwarding (if Windows guest)

### PR DESCRIPTION
**Pull request by bheuvel**
Closes #116

### CODE REVIEW

File: lib/vagrant-cloudstack/action/run_instance.rb [hunk 1]
```diff
@@ -193,49 +194,32 @@ def call(env)
...
-          if (pf_ip_address_id or pf_ip_address) and (pf_public_port or pf_public_port_randomrange)
-            if pf_private_port.nil?
-              communicator = env[:machine].communicate.instance_variable_get('@logger').instance_variable_get('@name')
-              comm_obj = env[:machine].config.send(communicator)
-
-              pf_private_port = comm_obj.port if comm_obj.respond_to?('port')
-              pf_private_port = comm_obj.guest_port if comm_obj.respond_to?('guest_port')
-              pf_private_port = comm_obj.default.port if (comm_obj.respond_to?('default') and comm_obj.default.respond_to?('port'))
-            end
+#         if pf_private_rdp_port.nil?
```
Comment (miguelaferreira) [inline lib/vagrant-cloudstack/action/run_instance.rb]: Did you forget to uncomment this line? Or should it be deleted?

(L200 too)

File: lib/vagrant-cloudstack/capabilities/rdp.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module VagrantPlugins
+  module Cloudstack
+    module Cap
```
Comment (miguelaferreira) [inline lib/vagrant-cloudstack/capabilities/rdp.rb:3]: does `module Cap` stand for module capability/capabilities?

**Comment by miguelaferreira** [comment]
When creating a box with kitchen, I see this output:

```
       ==> default: Forwarding ports...
           default: 3389 => 3389 (adapter 1)
           default: 5985 => 55985 (adapter 1)
           default: 5986 => 55986 (adapter 1)
```

What's up with the duplicated 5985 port forwarding message?
Duplicated log, or duplicated action?

EDIT: I'm still using virtualbox as a  provider, but I've pulled in vagrant from your fork.
`gem 'vagrant', git: 'https://github.com/bheuvel/vagrant.git', branch: 'rdpautologon'`

**Comment by miguelaferreira** [comment]
Still in virtualbox, every time I create a new VM with this branch I get this error:

```
==> default: Waiting for machine to boot. This may take a few minutes...
    default: WinRM address: 127.0.0.1:55985
    default: WinRM username: vagrant
    default: WinRM transport: plaintext
STDERR: An error occurred executing a remote WinRM command.

Shell: powershell
Command: hostname
if ($?) { exit 0 } else { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } }
Message: [WSMAN ERROR CODE: 2147942421]: <f:WSManFault Code='2147942421' Machine='127.0.0.1' xmlns:f='http://schemas.microsoft.com/wbem/wsman/1/wsmanfault'><f:Message><f:ProviderFault path='%systemroot%\system32\winrscmd.dll' provider='Shell cmd plugin'>The device is not ready. </f:ProviderFault></f:Message></f:WSManFault>
---- End output of vagrant up --no-provision --provider virtualbox ----
Ran vagrant up --no-provision --provider virtualbox returned 1]
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

The machine is created, and is functional. I can converge it and even RDP in, all via kitchen.

File: lib/vagrant-cloudstack/capabilities/rdp.rb (hunk 2 above)
Comment (bheuvel) [inline lib/vagrant-cloudstack/capabilities/rdp.rb:3]: Mostly monkey see, monkey do from other plugins and doc (http://docs.vagrantup.com/v2/plugins/guest-capabilities.html)
I assume 'Cap' module is used by Vagrant to list capabilities.

**Comment by bheuvel** [comment]
@miguelaferreira  Your "duplicated" port is actaully port 5985 AND 598_6_ (clear-text versus encrypted channel). Vagrant(-virtualbox) forwards both by default.

The failing WinRM is interesting to know if this happens every time, and if you then have this with vagrant on master. My vagrant branch is really only usefull if you are working from a Windows host, only in that case you will get auto-logon.

**Comment by miguelaferreira** [comment]
Ok. I will double check with vagrant from master.

**Comment by bheuvel** [comment]
It seems that the RDP portforwarding takes place for ALL guests. Need to make this for Windows only.

**Comment by miguelaferreira** [comment]
BTW via kitchen I'm not getting the RDP port forwarding and firewall rules. I will have to test directly with vagrant

**Comment by miguelaferreira** [comment]
This one did not work.

Gemfile:

``` ruby
source 'https://rubygems.org'

gemspec

group :development do
  # We depend on Vagrant for development, but we don't add it as a
  # gem dependency because we expect to be installed within the
  # Vagrant environment itself using `vagrant plugin`.
  gem 'vagrant', git: 'https://github.com/bheuvel/vagrant.git', branch: 'rdpautologon'
  gem 'coveralls', require: false
  gem 'simplecov', require: false
  gem 'rspec-core'
  gem 'rspec-expectations'
  gem 'rspec-its'
  gem 'rspec-mocks'
end

group :plugins do
  gem 'vagrant-cloudstack', path: '.'
end
```

Vagrantfile:

``` ruby
VAGRANTFILE_API_VERSION = '2'

Vagrant.require_version '>= 1.5.0'

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.vm.box = "..."

  config.vm.communicator = 'winrm'

  config.vm.synced_folder ".", "/vagrant", type: "rsync",
    rsync__exclude: [".git/", "vendor"],  disabled: true

  config.vm.provider :cloudstack do |cloudstack, override|
    cloudstack.host   = "..."
    cloudstack.path   = "/client/api"
    cloudstack.port   = "443"
    cloudstack.scheme = "https"
    cloudstack.api_key    = "..."
    cloudstack.secret_key = "..."
    cloudstack.expunge_on_destroy = false

    my_ip = "A.B.C.D"
    cloudstack.service_offering_name = "..."
    cloudstack.network_name          = "..."
    cloudstack.zone_name             = "..."
    cloudstack.network_type          = "Advanced"
    cloudstack.pf_ip_address         = my_ip
    cloudstack.pf_open_firewall      = true
    cloudstack.pf_trusted_networks   = "..." 
    cloudstack.vm_user = 'Administrator'
  end

end
```

Commands:

```
git fetch upstream pull/117/head:pr/117
git checkout pr/117
rm Gemfile.lock
bundle install
bundle exec vagrant up
bundle exec vagrant rdp
```

**Comment by bheuvel** [comment]
Perhaps explicitly add `config.vm.guest = :windows` to your Vagrantfile?

**Comment by miguelaferreira** [comment]
Yap, that makes a difference

```
    default: Syncronized resource: public_ip_address - ...
==> default:  -- IP address    : ...
==> default:  -- Protocol      : tcp
==> default:  -- Public port   : 65268
==> default:  -- Private port  : 5985
==> default:  -- Open Firewall : false
==> default: Creating a port forwarding rule for this instance ...
==> default: Fetching UUID for public_ip_address with ipaddress '...'
    default: Syncronized resource: public_ip_address - ...
==> default:  -- IP address    : ...
==> default:  -- Protocol      : tcp
==> default:  -- Public port   : 65405
==> default:  -- Private port  : 3389
==> default:  -- Open Firewall : false
```

**Comment by miguelaferreira** [comment]
And RDP works fine out of the box :+1: 

I will merge the PR. @bheuvel can you please update the README (or did I miss it?) with the details to make this work?

**Pull request merged**